observabilityclient/v1/deploy.py

```diff
@@ -0,0 +1,140 @@
+#   a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#   License for the specific language governing permissions and limitations
+#   under the License.
+#
+
+import os
+import requests
+import sys
+import yaml
+
+from osc_lib.i18n import _
+
+from observabilityclient.v1 import base
+from observabilityclient.utils import runner
+
+
+INVENTORY = os.path.join(base.OBSWRKDIR, 'openstack-inventory.yaml')
+ENDPOINTS = os.path.join(base.OBSWRKDIR, 'scrape-endpoints.yaml')
+STACKRC = os.path.join(base.OBSWRKDIR, 'stackrc')
+STACKRCKEYS = ('OS_AUTH_TYPE', 'OS_PASSWORD', 'OS_AUTH_URL', 'OS_USERNAME',
+               'OS_PROJECT_NAME', 'OS_NO_CACHE', 'COMPUTE_API_VERSION',
+               'OS_USER_DOMAIN_NAME', 'OS_CLOUDNAME', 'OS_PROJECT_DOMAIN_NAME',
+               'OS_IDENTITY_API_VERSION', 'NOVA_VERSION')
+
+
+def _curl(host: dict, port: int, timeout: int = 1) -> str:
+    """Returns scraping endpoint URL if it is reachable
+    otherwise returns None."""
+    url = f'http://{host["ip"]}:{port}/metrics'
+    try:
+        r = requests.get(url, timeout=1)
+    except requests.exceptions.ConnectionError:
+        url = None
+    if r.status_code != 200:
+        url = None
+    r.close()
+    return url
+
+
+class Discover(base.ObservabilityBaseCommand):
+    """Install and configure given Observability component(s)"""
+
+    def get_parser(self, prog_name):
+        parser = super().get_parser(prog_name)
+        parser.add_argument(
+            '--rcfile',
+            default=None,
+            help=_("Path to rc file used for Keystone authentication.")
+        )
+        return parser
+
+    def take_action(self, parsed_args):
+        # make sure we have rc file provided or ready
+        if not parsed_args.rcfile:
+            env = os.environ.keys()
+            have = True
+            for key in STACKRCKEYS:
+                if key not in env:
+                    have = False
+                    break
+            if not os.path.exists(STACKRC):
+                if have:
+                    with open(STACKRC, 'w') as f:
+                        for key in STACKRCKEYS:
+                            f.write('{}={}\n'.format(key, os.environ[key]))
+                    os.chmod(STACKRC, 0o600)
+                else:
+                    print(_('Keystone auth is required. Either provide rc file'
+                            'or source one before running the command.'))
+                    sys.exit(1)
+
+        # discover undercloud and overcloud nodes
+        stackrc = parsed_args.rcfile if parsed_args.rcfile else STACKRC
+        rc, out, err = self._execute(
+            '. {} && tripleo-ansible-inventory '
```

Comment:
If this normal for openstackclient plugins? I'm used to the idea that I source the correct stackrc and THEN I run openstack commands. It's unexpected for me to find code in the plugin that sources the stackrc. This isn't authoritative, I'm just surprised and wondering if this is a good idea or not.

Reply:
Yes the idea of yours is correct. I've implemented that for the case where you will want to script observability commands. Maybe I overengineered and remove that for now and coma back to it when needed.